test/remote/gateways/remote_square_test.rb

```diff
@@ -0,0 +1,395 @@
+    @options = {
+      billing_address: address,
+      description: 'Store Purchase Note'
+    }
+  end
+
+  def test_only_accepts_card_nonce_not_creditcard_pan
+    credit_card_pan = '4111111111111111' # should be rejected
+    response = @gateway.purchase(@amount, credit_card_pan, @options)
+    assert_failure response
+    assert_not_nil err = response.params['errors'].first
+    assert_equal 'INVALID_REQUEST_ERROR', err['category']
+    assert_equal 'NOT_FOUND', err['code']
+    assert_equal 'card_nonce', err['field']
+    assert_nil response.error_code
+  end
+
+  def test_successful_purchase
+    response = @gateway.purchase(@amount, @credit_card, @options)
+    assert_success response
+    assert_equal 1, response.responses.count
+    assert_equal 'Success', response.message
+    assert_nil response.error_code
+  end
+
+  def test_successful_purchase_with_more_options
+    options = {
+      idemepotency_key: SecureRandom.uuid,
+      shipping_address: address,
+      billing_address: address,
+      buyer_email_address: "joe@example.com",
+      order_id: 'OrderNum123',
+      description: 'custom description note you ordered xyz'
+    }
+
+    response = @gateway.purchase(@amount, @credit_card, options)
+
+    assert_success response
+    assert_not_nil txn = response.params['transaction']
+    assert_equal 'OrderNum123', txn['reference_id']
+    assert_equal 'CAPTURED', txn['tenders'].first['card_details']['status']
+    assert_equal 'custom description note you ordered xyz', txn['tenders'].first['note']
+    assert_equal 'Success', response.message
+    assert_nil response.error_code
+end
+
+  # General purchase and authorize tests.
+
+  def test_failed_purchase_decline
+    response = @gateway.purchase(@amount, @declined_card, @options)
+    assert_failure response
+    assert_equal 1, response.responses.count
+    assert_equal 'Card declined.', response.message
+    assert_equal 'card_declined', response.error_code
+  end
+
+  def test_failed_purchase_invalid_cvv
+    response = @gateway.purchase(@amount, 'fake-card-nonce-rejected-cvv', @options)
+    assert_failure response
+    assert_equal 'Card verification code check failed.', response.message
+    assert_equal 'incorrect_cvc', response.error_code
+  end
+
+  def test_failed_purchase_rejected_avs_zip
+    response = @gateway.purchase(@amount, 'fake-card-nonce-rejected-postalcode', @options)
+    assert_failure response
+    assert_equal 'Postal code check failed.', response.message
+    assert_equal 'incorrect_zip', response.error_code
+  end
+
+  def test_failed_purchase_expiration_incorrect
+    response = @gateway.purchase(@amount, 'fake-card-nonce-rejected-expiration', @options)
+    assert_failure response
+    assert_equal 'Invalid card expiration date.', response.message
+    assert_equal 'invalid_expiry_date', response.error_code
+  end
+
+  def test_successful_authorize_and_capture
+    @options.merge!({delay_capture: true})
```

Comment:
I'm not that familiar with `TestUnit`, but just to make sure, `@options` is reset before each test, right?